Adding typealias and automatically releasing objects

cstjean opened the pull request:
I'd like to add GIF support this week, and I wanted to clean up the code a bit beforehand. I'm not sure I like the typealias changes. It looks noisier, but it's probably for the best?

Also, `save_and_release` never actually released the image, so there was a memory leak, now fixed.

rsrock commented on src/QuartzImageIO.jl:
```diff
@@ -246,14 +280,14 @@ function save_(fname, img::Image, image_type)
+                do_save(cgImage, fname, image_type)
+            end
+        end                           
+    end
```
I don't think I like this `with_obj` approach. The block nesting can quickly get out of hand.  I also find the original code much more readable.

cstjean replied:
I agree, but isn't the solution with `try` + `finally` even worse? Is there anything cleaner? Or do we simply accept that memory leaks will happen when exceptions are thrown?

rsrock commented:
Thanks for working on this, sorry that it's taken so long to review.  I like the typealiases, but I'm not so keen on the `with_obj` (see the inline comments).  Could you revert those changes?  If it's easier, you can address the memory leak in a separate PR.

For error handling, I'd like to handle the NULLs and other errors at the call site, like in http://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/ under the `getenv()` example.   The error handling in the loading code could use some love too, along these lines.

rsrock commented on src/QuartzImageIO.jl:
```diff
@@ -222,14 +255,15 @@ end
-    CGImageDestinationAddImage(out_dest, cg_img);
-    CGImageDestinationFinalize(out_dest);
-    CFRelease(out_dest)
-    CFRelease(out_url)
```
Wouldn't a CFRelease(cg_img) do the trick here?

Also, we can get rid of the semicolons above

cstjean replied:
Will do.

cstjean commented:
I reverted the last two commits and fixed the leak. If you would like, I can squash it as is and we can merge it ready for a tagged release.

As for the error handling code, I'm happy to either leave it to you, or do it myself - if we can agree about what needs to be done. I like the `getenv` example and I can change the current `check_null` calls for that approach. 

For the release of objects, I would use `try` `finally`, but that will also lead to "deep nesting", unless we factor out some of the code into smaller methods. To me that's still better than the current error handling with `if`s (I can see at least two missing `CFRelease(dict)` statements in the current loading code).

Finally, it seems that we should ideally bail out to ImageMagick on every error. I would wrap the whole loading code into something like

```julia
function load(filename)
   try
      load_impl(filename)
   catch
      if debug_flag
          rethrow()
      else
          return nothing # bail out to ImageMagick
      end
   end
end
```

timholy commented:
Bump. Should I merge this?

cstjean commented:
I'm fine with merging it. Should I squash first?

timholy commented:
If you wouldn't mind, that would be great. (No need to squash to a single commit, unless you want to, but at least squashing out the "revert"s would be good.)

timholy commented:
Also hoping @rsrock or someone can review, but if he's too busy I'll merge by Monday or so.

timholy commented:
I understand that @rsrock is too busy, so I'll merge whenever you squash.

cstjean commented:
Done.

timholy commented:
Thanks!
